agent/job_runner.go

```diff
@@ -587,6 +621,34 @@ func truncateEnv(l logger.Logger, env map[string]string, key string, max int) er
 	return nil
 }
 
+func (r *JobRunner)executePreBootstrapHook(hook string) (bool, error) {
+	sh, err := shell.New()
+	if err != nil {
+		return false, err
+	}
+
+	// TODO stream logging up to buildkite?
+	// TODO if capturing and printing output of this script do we expect the redactors to run?
+
+	// TODO do we expect this script to be executed by bash, or sourced by bash
```

Comment:
I think we should introduce a new terminology to describe these new 'hooks' that aren’t part of the bootstrap process and are exec'd instead of sourced.

Lifecycle Hooks?

Reply:
Another benefit of these 'Agent Lifecycle Hooks' being exec'd and not sourced by Bash is that they can be written in a higher level language. I think it’s reasonable to expect this one in particular to be written in something like golang, or Rust and placed in the right spot for the buildkite-agent to fork/exec.

Reply:
I like the idea of exec-ing them over sourcing them for the reasons you describe! Also, as per my comment about STDOUT and sending the log back to BK, I think it's OK for agent lifecycle hooks to have different behaviours. I tried to think of another word than "hook" though... We already have plugins... 🤔